detection_rules/ghwrap.py

```diff
@@ -0,0 +1,313 @@
+        sha256 = hashlib.sha256(contents).hexdigest()
+
+        assets[zipped.filename] = {
+            'contents': contents,
+            'metadata': {
+                'compress_size': zipped.compress_size,
+                # zipfile provides only a 6 tuple datetime; -1 means DST is unknown;  0's set tm_wday and tm_yday
+                'created_at': time.strftime('%Y-%m-%dT%H:%M:%SZ', zipped.date_time + (0, 0, -1)),
+                'sha256': sha256,
+                'size': zipped.file_size,
+            }
+        }
+
+    return zipped_sha256, assets
+
+
+def load_json_gh_asset(url: str) -> dict:
+    """Load and return the contents of a json asset file."""
+    response = requests.get(url)
+    response.raise_for_status()
+    return response.json()
+
+
+def download_gh_asset(url: str, path: str, overwrite=False):
+    """Download and unzip a GitHub asset."""
+    zipped = requests.get(url)
+    z = ZipFile(io.BytesIO(zipped.content))
+
+    Path(path).mkdir(exist_ok=True)
+    if overwrite:
+        shutil.rmtree(path, ignore_errors=True)
+
+    z.extractall(path)
+    click.echo(f'files saved to {path}')
+
+    z.close()
+
+
+class GithubClient:
+    """GitHub client wrapper."""
+
+    def __init__(self, token: Optional[str] = None):
+        """Get an unauthenticated client, verified authenticated client, or a default client."""
+        if not Github:
+            raise ModuleNotFoundError('Missing PyGithub - try running `pip install -r requirements-dev.txt`')
+
+        self.client: Github = Github(token)
+        self.unauthenticated_client = Github()
+        self.__token = token
+        self.__authenticated_client = None
+
+    @property
+    def authenticated_client(self) -> Github:
+        if not self.__token:
+            raise ValueError('Token not defined! Re-instantiate with a token or use add_token method')
+        if not self.__authenticated_client:
+            self.__authenticated_client = Github(self.__token)
+        return self.__authenticated_client
+
+    def add_token(self, token):
+        self.__token = token
+
+
+@dataclass
+class AssetManifestEntry:
+
+    compress_size: int
+    created_at: datetime
+    name: str
+    sha256: str
+    size: int
+
+
+@dataclass
+class AssetManifestMetadata:
+
+    relative_url: str
+    entries: Dict[str, AssetManifestEntry]
+    zipped_sha256: definitions.Sha256
+    created_at: datetime = field(default_factory=datetime.utcnow)
+    description: Optional[str] = None  # populated by GitHub release asset label
+
+
+@dataclass
+class ReleaseManifest:
+
+    assets: Dict[str, AssetManifestMetadata]
+    assets_url: str
+    author: str  # parsed from GitHub release metadata as: author[login]
+    created_at: str
+    html_url: str
+    id: int
+    name: str
+    published_at: str
+    url: str
+    zipball_url: str
+    tag_name: str = None
+    description: str = None  # parsed from GitHub release metadata as: body
+
+
+class Manifest:
```

Comment:
might be good to give this a rename. ManifestManager or something. makes it one step less confusing to see ReleaseManifest and Manifest